clean up enum handling

hkiel opened the pull request:
### Purpose

Trying to find a reason for false assert failure
```
LOG_INIT          | info    | ### START INITIALIZATION ###
assert            | warning | The following assertion has been violated at time 0.000000
|                 | |       | e1 >= initializationTests.E.E1 and e1 <= initializationTests.E.En
assert            | warning | Variable violating min/max constraint: initializationTests.E.E1 <= e1 <= initializationTests.E.En, has value: 0
```
in `simulation/modelica/initialization/parameters.mo`

Variable `e1` is using default value (`initializationTests.E.E1`) and should not produce an assert failure.

perost commented:
A suggestion would be to add a function for getting the n:th enumeration literal (in `Types.mo` probably) in order to simplify the code a bit, like:
```Modelica
public function getNthEnumLiteral
  "Returns the n:th literal of an enumeration type."
  input DAE.Type ty;
  input Integer n;
  output DAE.Exp literalExp;
algorithm
  literalExp := match ty
    case DAE.T_ENUMERATION()
      then DAE.ENUM_LITERAL(AbsynUtil.suffixPath(ty.path, listGet(ty.names, n)), n);

    // Not sure if this is correct, maybe it would be more correct to 
    // use the path in the ClassInf.State inside the subtype?
    case DAE.T_SUBTYPE_BASIC()
      then getNthEnumLiteral(ty.complexType, n);
  end match;
end getNthEnumLiteral;
```
I haven't actually tested this code, so beware. But I think it's something we should have added a long time ago.

hkiel commented:
@perost In your suggestion you use `AbsynUtil.suffixPath()` to create the default variable, whereas currently we use `AbsynUtil.joinPaths()`. I see that the enum identifier is just a suffix to the enumeration, but the result looks different:
```
-[<interactive>:10:5-10:19:writable] Warning: Parameter e1 has no value, and is fixed during initialization (fixed=true), using available start value (start=initializationTests.E.E1) as default value.
+[<interactive>:10:5-10:19:writable] Warning: Parameter e1 has no value, and is fixed during initialization (fixed=true), using available start value (start=.initializationTests.E.E1) as default value.
```

suffixPath() prepends a leading dot to the path, whereas when you explicitly state the start value  (e.g. `parameter E e3 (start=E.E3);`) there is also no dot at the beginning.
I would go for the `joinPaths()` version for now as it looks more consistent.

perost commented:
> @perost In your suggestion you use `AbsynUtil.suffixPath()` to create the default variable, whereas currently we use `AbsynUtil.joinPaths()`. I see that the enum identifier is just a suffix to the enumeration, but the result looks different:
> 
> ```
> -[<interactive>:10:5-10:19:writable] Warning: Parameter e1 has no value, and is fixed during initialization (fixed=true), using available start value (start=initializationTests.E.E1) as default value.
> +[<interactive>:10:5-10:19:writable] Warning: Parameter e1 has no value, and is fixed during initialization (fixed=true), using available start value (start=.initializationTests.E.E1) as default value.
> ```
> 
> suffixPath() prepends a leading dot to the path, whereas when you explicitly state the start value (e.g. `parameter E e3 (start=E.E3);`) there is also no dot at the beginning. I would go for the `joinPaths()` version for now as it looks more consistent.

It's actually the opposite, `joinPaths` removes the leading dot from the path while `suffixPath` preserves it. I used `suffixPath` since it's a bit more efficient and I assumed both functions would do the sane thing and preserve the qualifiedness of the path, but I should know better than to expect sanity from the old frontend :smile:

So sure, use `joinPaths` instead since it gives the expected result.

hkiel commented:
```
LOG_INIT_V        | info    | parameter values
|                 | |       | | real parameters
|                 | |       | | | [1] parameter Real r1(start=0, fixed=true) = 0
|                 | |       | | | [2] parameter Real r2(start=2, fixed=false) = 2
|                 | |       | | | [3] parameter Real r3(start=3, fixed=true) = 3
|                 | |       | | integer parameters
|                 | |       | | | [1] parameter Integer e1(start=0, fixed=true) = 0
|                 | |       | | | [2] parameter Integer e2(start=2, fixed=false) = 2
|                 | |       | | | [3] parameter Integer e3(start=3, fixed=true) = 3
|                 | |       | | | [4] parameter Integer i1(start=0, fixed=true) = 0
|                 | |       | | | [5] parameter Integer i2(start=2, fixed=false) = 2
|                 | |       | | | [6] parameter Integer i3(start=3, fixed=true) = 3
```

Looks like enums are handled too much like integers. Ints default to zero, but enums **should** default to one.

hkiel commented on OMCompiler/Compiler/BackEnd/Initialization.mo:
```diff
@@ -917,11 +917,7 @@ algorithm
     Error.addSourceMessage(Error.UNBOUND_PARAMETER_WITH_START_VALUE_WARNING, {s, str}, info);
```
The line above look like they would add the default start value, but it is not used in SimCode. An enum parameter without start value is initialized as an integer with a zero value.
Anyone got an idea why this happens?

perost reviewed (approved):
Looks good to me, though I don't really know much about how the backend handles start values.

lochel reviewed (approved):
Looks really good as far as I can tell 👍

hkiel commented:
Can you merge please?
